### PR TITLE
Stable null-safety release

### PR DESCRIPTION
**Pull request by wolfenrain**
I had to disable `flutter_driver` in the example as it is depended on `crypto`, but as audioplayers is also dependend on `crypto` through `uuid`, it could not resolve the proper version.

Related to https://github.com/luanpotter/audioplayers/issues/784

### CODE REVIEW

**Comment by deimantasa** [comment]
Hi @wolfenrain ,

Since I've created a FR post, I've been also working on `null-safety` under the hood. Problem with `crypto` can be resolved. All you need is to include 
```
dependency_overrides:
  crypto: ^3.0.0
```
to example's `pubspec.yaml`. Then everything should compile well.


Also, have you tried running the demo app? After migration, there seems to be tons of small issues, which needs to be fixed manually. We cannot fully rely on `dart migration`.

Cheers

**Comment by wolfenrain** [comment]
I didn't use dart migrate as it is unreliable when it comes to more complex code, I prefer to follow the context of the code and do it manually (which I did in this case).

I'll have a look at your crypto solutions tomorrow, I didn't expect that to solve the problem with flutter driver, but I suppose the version mayor increase was not truly breaking, just null-safety breaking

**Comment by deimantasa** [comment]
Sure. If you're interested - have a look [here](https://github.com/deimantasa/audioplayers/tree/feature/null-safety-migration). It's nearly done - just couple of small fixes needs to be made I think. Commits are very separated so it's easy to understand when what change was done.

Cheers

**Comment by b-jan** [comment]
Hello, do you plan to release it soon?

**Comment by deimantasa** [comment]
> I tried as you said, but it still wont run the drivers, did you add another thing to make them work? I already discussed with @luanpotter that we can temporary disable them. So it is not a must for the release, but would be nice to have.

I had a quick look - your's and my refactor are quite different. Not sure why is that. But anyways, as long as it works and everyone agrees upon the solution.

Cheers

**Comment by wolfenrain** [comment]
> > I tried as you said, but it still wont run the drivers, did you add another thing to make them work? I already discussed with @luanpotter that we can temporary disable them. So it is not a must for the release, but would be nice to have.
> 
> I had a quick look - your's and my refactor are quite different. Not sure why is that. But anyways, as long as it works and everyone agrees upon the solution.
> 
> Cheers

Might be because I did not change it using `dart migrate`, instead I updated the code based on the context, for instance a lot of parameters in your code are nullable, and inside the function those parameters get set if `null`, I made those parameters non-nullable with a default value that was already inside that function, but removed the code that set it again.

So I did more then just setting the proper `?`, I also updated the code to be more inline with proper null-safety APIs.

**Review by Dimoshka — DISMISSED** [review]

File: lib/audioplayers.dart [hunk 1]
```diff
@@ -308,30 +308,27 @@ class AudioPlayer {
   PlayerMode mode;
 
   /// Creates a new instance and assigns an unique id to it.
-  AudioPlayer({this.mode = PlayerMode.MEDIA_PLAYER, this.playerId}) {
-    this.mode ??= PlayerMode.MEDIA_PLAYER;
-    this.playerId ??= _uuid.v4();
+  AudioPlayer({this.mode = PlayerMode.MEDIA_PLAYER, this.playerId = ''}) {
+    this.playerId = this.playerId == '' ? _uuid.v4() : this.playerId;
```
Comment (luanpotter) [inline lib/audioplayers.dart:312]: should we use `null` as the default value for player id?
```
  AudioPlayer({this.mode = PlayerMode.MEDIA_PLAYER, String? playerId}) {
    this.playerId = playerId == ?? _uuid.v4();
```

File: lib/audioplayers.dart [hunk 2]
```diff
@@ -249,7 +249,7 @@ class AudioPlayer {
 
   /// Handler of changes on player state.
   @deprecated
```
Comment (luanpotter) [inline lib/audioplayers.dart:251]: can be a followup, but we should use this opportunity to clean this up

File: lib/audioplayers.dart [hunk 3]
```diff
@@ -551,10 +538,10 @@ class AudioPlayer {
   ///
   /// Specify atleast title
   Future<dynamic> setNotification({
-    String title,
-    String albumTitle,
-    String artist,
-    String imageUrl,
+    String title = '',
```
Comment (luanpotter) [inline lib/audioplayers.dart:541]: maybe these should be nullable? gotta check on the ios side what happens, I can imagine they should be optional

File: lib/audioplayers.dart (hunk 2 above)
Comment (wolfenrain) [reply lib/audioplayers.dart:251]: I decided to do it in this PR as it is barely any code change

File: lib/audioplayers.dart (hunk 3 above)
Comment (wolfenrain) [reply lib/audioplayers.dart:541]:
```swift
    func setNotification(
        playerId: String,
        title: String?,
        albumTitle: String?,
        artist: String?,
        imageUrl: String?,
        forwardSkipInterval: Int,
        backwardSkipInterval: Int,
        duration: Int?,
        elapsedTime: Int,
        enablePreviousTrackButton: Bool?,
        enableNextTrackButton: Bool?
    ) {
```
They are all nullables on the swift it seems.

File: lib/audioplayers.dart (hunk 1 above)
Comment (wolfenrain) [reply lib/audioplayers.dart:312]: As far as I can see, the iOS side expects a non-nullable String

**Comment by wolfenrain** [comment]
I haven't yet tested this version of the branch yet. But I followed up on the reviews, and removed the deprecated code

**Review by erickzanardo-dx — APPROVED** [review]

**Review by erickzanardo — APPROVED** [review]

**Comment by BringTheChill** [comment]
It does not work to change volume on iOS, I get: `"Error calling setVolume, volume cannot be null"`

**Comment by wolfenrain** [comment]
> setVolume

@BringTheChill are you by chance passing a `double?`, because the `setVolume` on the Dart layer is already non-nullable.

**Comment by Zeswen** [comment]
Great work with null-safety! When are you going to create a new release with it @luanpotter?

**Comment by BringTheChill** [comment]
@wolfenrain I pass just non-nullable values

**Comment by luanpotter** [comment]
Will launch it later today, I still need to test it

**Pull request merged**